### PR TITLE
Eliminate Spaceship false-positive

### PR DESCRIPTION
**Pull request by sidz**
This PR eliminates Spaceship false-positive submitted in https://github.com/infection/infection/issues/1438.

- [x] Covered by tests

### CODE REVIEW

File: tests/phpunit/Mutator/Sort/SpaceshipTest.php [hunk 1]
```diff
@@ -70,5 +70,53 @@ public function mutationsProvider(): iterable
...
+
+($a <=> $b) === 0;
+PHP
+        ];
+
+        yield 'It does not swap operators when result is equal to zero on the right side' => [
+            <<<'PHP'
+<?php
+
+($a <=> $b) == 0;
```
Comment (sanmai) [inline tests/phpunit/Mutator/Sort/SpaceshipTest.php:86]: How about `($a <=> $b) == 0.0;`?

Comment (sidz) [reply tests/phpunit/Mutator/Sort/SpaceshipTest.php:86]: will add.

I thought about this case but assumed that people won't use floats here :)

Comment (sidz) [reply tests/phpunit/Mutator/Sort/SpaceshipTest.php]: @sanmai done

File: src/Mutator/Sort/Spaceship.php [hunk 2]
```diff
@@ -74,6 +74,65 @@ public function mutate(Node $node): iterable
...
+            }
+
+            if ($parentAttribute->left instanceof Node\Scalar\DNumber && $parentAttribute->left->value === 0.0) {
+                return true;
+            }
+
+            if (
+                $parentAttribute->left instanceof Node\Scalar\String_
+                && is_numeric($parentAttribute->left->value)
+                && ($parentAttribute->left->value === '0' || $parentAttribute->left->value === '0.0')
```
Comment (sanmai) [inline src/Mutator/Sort/Spaceship.php:130]: Do you this `0 === (int) $parentAttribute->left->value` won't work?

File: tests/phpunit/Mutator/Sort/SpaceshipTest.php (hunk 1 above)
Comment (sanmai) [reply tests/phpunit/Mutator/Sort/SpaceshipTest.php]: People do things :(

File: src/Mutator/Sort/Spaceship.php (hunk 2 above)
Comment (sidz) [reply src/Mutator/Sort/Spaceship.php:130]: what if somebody decided to compare with 0.1 ? :)

**Review by sanmai — APPROVED** [review]

**Comment by sidz** [comment]
thanks @sanmai

File: src/Mutator/Sort/Spaceship.php [hunk 3]
```diff
@@ -74,6 +74,65 @@ public function mutate(Node $node): iterable
...
+                return true;
+            }
+
+            if ($parentAttribute->right instanceof Node\Scalar\DNumber && $parentAttribute->right->value === 0.0) {
+                return true;
+            }
+
+            if (
+                $parentAttribute->right instanceof Node\Scalar\String_
+                && is_numeric($parentAttribute->right->value)
```
Comment (theofidry) [inline src/Mutator/Sort/Spaceship.php:113]: missing the `is_numeric` import

File: src/Mutator/Sort/Spaceship.php [hunk 4]
```diff
@@ -74,6 +74,65 @@ public function mutate(Node $node): iterable
...
+        $parentAttribute = $node->getAttribute('parent');
+
+        if ($parentAttribute instanceof Node\Expr\BinaryOp\Identical) {
+            if ($parentAttribute->right instanceof Node\Scalar\LNumber && $parentAttribute->right->value === 0) {
+                return true;
+            }
+
+            if ($parentAttribute->left instanceof Node\Scalar\LNumber && $parentAttribute->left->value === 0) {
+                return true;
+            }
```
Comment (theofidry) [inline src/Mutator/Sort/Spaceship.php:99]: maybe you can return `false` here already?

What do you think of extracting the content of each `if()` in a private method for better readability?

**Review by theofidry — COMMENTED** [review]
👍

**Pull request merged**